T/224

pjasiun opened the pull request:
Fixes https://github.com/ckeditor/ckeditor5-core/issues/224 with some additional clean-up.

scofalik commented on src/treemodel/operation/attributeoperation.js:
```diff
@@ -90,19 +90,19 @@ export default class AttributeOperation extends Operation {
 				 *
 				 * @error operation-attribute-no-attr-to-remove
-				 * @param {core.treeModel.Node} node
+				 * @param {core.treeModel.Node|core.treeModel.TextFragment} item
```
This can be `core.treeModel.Item` now

pjasiun replied:
Fixed.

scofalik commented on src/treemodel/position.js:
```diff
@@ -411,6 +411,8 @@ export default class Position {
 	/**
 	 * Creates a new position after given node.
 	 *
+	 * @see core.treeModel.TreeWalkerValue
```
Are you sure this is the correct way to use @see tag? I am not sure either and have seen various way to use it.

Reinmar replied:
Run `gulp docs` and you'll know :D.

pjasiun replied:
Fixed.

scofalik commented on src/treemodel/range.js:
```diff
@@ -261,32 +252,51 @@ export default class Range {
 	/**
-	 * Returns an iterator that iterates over all {@link core.treeModel.Node nodes} that are in this range and returns them.
+	 * Returns an iterator that iterates over all {@link core.treeModel.Item items} that are in this range and returns
+	 * them together together with additional information like length or {@link core.treeModel.Position positions},
```
"together together"

pjasiun replied:
Fixed.

scofalik commented on src/treemodel/range.js:
```diff
@@ -261,32 +252,51 @@ export default class Range {
-	 * **Note:** this method will not return a parent node of start position. This is in contrary to {@link core.treeModel.TreeWalker}
-	 * which will return that node with `'ELEMENT_END'` type. This method also returns each {@link core.treeModel.Element} once,
-	 * while simply used {@link core.treeModel.TreeWalker} might return it twice: for `'ELEMENT_START'` and `'ELEMENT_END'`.
+	 * **Node:*** because this method does not return {@link core.treeModel.TreeWalkerValue values} with the type of
```
Node :)

pjasiun replied:
Fixed.

scofalik commented on src/treemodel/range.js:
```diff
@@ -261,32 +252,51 @@ export default class Range {
-	*getAllNodes( mergeCharacters ) {
-		let it = new TreeWalker( { boundaries: this, mergeCharacters: mergeCharacters } );
-		let step;
+	*getValues( mergeCharacters ) {
```
I am not sure if I like the name of the method, despite it returning an object which class has "Value" in name.

It feels strange to ask about "range values". What values range has? Range might contain nodes or items, but not values.

Reinmar replied:
+1

scofalik replied:
After reading tests I am not sure whether someone would even like to use `getValues` if they have `getItems` which essentialy does very similar things, but is easier to operate with.

pjasiun replied:
So could you suggest a better name?

scofalik replied:
Maybe this should not have a name but just be a default range iterator?

pjasiun replied:
The whole point of this change is to return `position` and `length` together with the `item`. I have no better idea to call the object will contains such a group of data.

pjasiun replied:
Maybe `getData` and `TreeWalkerData`?

scofalik replied:
AFAIR we do not use range iterator all that often because it was cumbersome.

pjasiun replied:
I removed default iterator, because we do not know if one what to `mergeCharacters` or not and if one want to return `ELEMENT_END` or not. I agree it was useless with the form it was. For me we could make this method default iterator with the `mergeCharacters`, so it will skip `ELEMENT_END` and merge characters. It is the most convenient for me, but I do not know if it is what people expect.

scofalik replied:
The most common usage will be either this or `getItems`. If someone want to "go deep", they always have the possibility to use `TreeWalker` and set boundaries to given range or use `getPositions`. Cases where you just want to see what items are in range are the most common.

scofalik commented on src/treemodel/treewalker.js:
```diff
@@ -238,13 +250,23 @@ function formatReturnValue( type, item ) {
- * @property {treeModel.Node|treeModel.TextFragment} item Value between old and new position of {@link core.treeModel.TreeWalker}.
+ * @typedef {Object} core.treeModel.TreeWalkerValue
+ * @property {treeModel.TreeWalkerValueType} type
+ * @property {core.treeModel.Item} item Item between previous and next positions of {@link core.treeModel.TreeWalker}.
```
I don't like "between previous and next positions of TreeWalker". Old and new seems more appropriate and doesn't care whether we go `.next()` or `.previous()`. Alternative description might be `Item encountered during TreeWalker iteration` or smth like this.

pjasiun replied:
Fixed.

scofalik commented on src/treemodel/treewalker.js:
```diff
@@ -238,13 +250,23 @@ function formatReturnValue( type, item ) {
+ * @typedef {Object} core.treeModel.TreeWalkerValue
+ * @property {treeModel.TreeWalkerValueType} type
+ * @property {core.treeModel.Item} item Item between previous and next positions of {@link core.treeModel.TreeWalker}.
+ * @property {treeModel.Position} previousPosition Previous position of the iterator. For `'ELEMENT_END'` it is the last
```
Please update to `core.treeModel.Position`

pjasiun replied:
Fixed, here and is some other places.

scofalik commented on src/treemodel/treewalker.js:
```diff
@@ -238,13 +250,23 @@ function formatReturnValue( type, item ) {
+ * {@link core.treeModel.Position.createBefore}. It is also more efficient to get the position after node by shifting
+ * `previousPosition` by `length`, using {@link core.treeModel.Position#getShiftedBy}, then calculate the position using
+ * {@link core.treeModel.Position.createAfter}.
+ * @property {treeModel.Position} nextPosition Next position of the iterator. For `'ELEMENT_START'` it is the first
```
Please update to `core.treeModel.Position`

pjasiun replied:
Fixed.

pjasiun commented on src/treemodel/range.js:
```diff
@@ -261,32 +252,51 @@ export default class Range {
+	 * @returns {Iterable.<core.treeModel.Item>}
+	 */
+	*getItems( mergeCharacters ) {
+		for ( let value of this.getValues( mergeCharacters ) ) {
```
In the F2F discussion we decided that `mergeCharacters` will be usually true so should be changed to `singleCharacters`, which is false by default.

pjasiun commented:
I have changed default `TreeWalker` option to merge characters, since this is the way we usually use this class. The change is pretty big, because it means that we prefer `TextFragment` over `CharacterProxy` right now, but it also looks like a change in the good direction. I also removed the `Range.getValues` method and added a iterable interface to the `Range` which now returns merged values without `ELEMENT_END`. It looks like a most useful structure.

scofalik commented:
I've pushed some minor fixes.
